lib/fluent/plugin/kubernetes_metadata_watch_pods.rb

```diff
@@ -70,6 +92,14 @@ def start_pod_watch
             @stats.bump(:pod_cache_watch_ignored)
         end
       end
+    rescue Exception => e
+      # Instead of raising exceptions and crashing Fluentd, swallow the
+      # exception and reset the watcher.
+      log.info "Exception encountered parsing pod watch event. " \
+               "The connection might have been closed. " \
+               "Resetting the pod watcher.", e
+      Thread.current[:pod_watcher_alive] = false
```

Comment:
@jcantrill, Ling and I had a discussion offline, here's our takeaways:

Adding the two metrics below would allow us to approximate downtime for API servers.
- retry_sleep_total metric to sum up time that we're sleeping
- retry_attempts metric to sum up the number of reconnects

We could consider allowing the original exception to be thrown if the time slept exceeds a critical threshold.

Reply:
> @jcantrill, Ling and I had a discussion offline, here's our takeaways:
> 
> Adding the two metrics below would allow us to approximate downtime for API servers.
> 
>     * retry_sleep_total metric to sum up time that we're sleeping
> 
>     * retry_attempts metric to sum up the number of reconnects

WFM.  The metrics that exist are not exposed to anything really consumable since it was part of the initial caching impl to get some details about what was going on.  Future task is likely to make the prometheus consumable.

> 
> 
> We could consider allowing the original exception to be thrown if the time slept exceeds a critical threshold.

In the past where I have utilized an exponential backoff, I have simply started the backoff loop over.  I don't know if there is value in throwing the exception but there is at least value in logging it every so many times we see it.

Reply:
Added two metrics: pod_watch_failures and namespace_watch_failures

Added retry_backoff and retry_max_times. I changed it to log an error for every 10 retries (roughly 17 minutes) and raise an exception to let fluentd restart. This way it something is completely wrong, a Fluentd restart could hopefully get it fixed.